Initial KeyVaultBackupClient for Java (#12671)
Azure/azure-sdk-for-java: sdk/keyvault/azure-security-keyvault-administration/README.md

# Azure Key Vault Administration library for Java

## Getting started
### Adding the package to your project

> Maven dependency for the Azure Key Vault Administration client library. Add it to your project's POM file.

Reviewer: There are multiple clients, so probably just drop "client" and say "... Administration library."

    ```

### Authenticate the client

> In order to interact with the Azure Key Vault service, you'll need to create an instance of the [KeyClient](#create-backup-client) class. You would need a **vault url** and **client secret credentials (client id, client secret, tenant id)** to instantiate a client object using the default `DefaultAzureCredential` examples shown in this document.

Reviewer 2: ```suggestion
In order to interact with the Azure Key Vault service, you'll need to create an instance of the [KeyVaultBackupClient](#create-backup-client) class. You would need a **vault url** and **client secret credentials (client id, client secret, tenant id)** to instantiate a client object using the default `DefaultAzureCredential` examples shown in this document.
```

    ```

* Take note of the service principal objectId
    ```Bash
    az ad sp show --id <appId> --query objectId
    ```
  
    Output:
    ```
    "<your-service-principal-object-id>"
    ```

* Use the aforementioned Key Vault name to retrieve details of your Key Vault, which also contain your Key Vault URL:

    ```Bash
    az keyvault show --name <your-key-vault-name>
    ```

#### Create Backup client
Once you've populated the **AZURE_CLIENT_ID**, **AZURE_CLIENT_SECRET**, and **AZURE_TENANT_ID** environment variables and replaced **your-key-vault-url** with the URI returned above, you can create the KeyVaultBackupClient:

```Java
import com.azure.identity.DefaultAzureCredentialBuilder;
import com.azure.security.keyvault.administration.KeyVaultBackupClient;
import com.azure.security.keyvault.administration.KeyVaultBackupClientBuilder;

KeyVaultBackupClient backupClient = new KeyVaultBackupClientBuilder()
    .vaultUrl("<your-key-vault-url>")
    .credential(new DefaultAzureCredentialBuilder().build())
    .buildClient();
```

> NOTE: For using an asynchronous client use KeyVaultBackupAsyncClient instead of KeyVaultBackupClient and call `buildAsyncClient()`

## Key concepts

> ### Key Vault Access Control client

Reviewer 2: ```suggestion
### Key Vault Backup client
```

### Restore Operation
A restore operation represents a long running operation for both a full key and selective key restore.

## Examples
### Sync API
The following sections provide several code snippets covering some of the most common Azure Key Vault Backup service tasks, including:
- [Backup a Key Vault](#backup-a-key-vault)
- [Restore a Key Vault](#restore-a-collection-of-keys)
- [Restore a key](#selectively-restore-a-key)

### Backup a Key Vault
Back up an entire collection of keys using `beginBackup`.

```java
SyncPoller<KeyVaultBackupOperation, Void> backupPoller =
    backupClient.beginBackup("<blob-storage-url-to-save-backup>", "<sas-token>");

backupPoller.waitForCompletion();

System.out.printf("Backup created");
```

### Restore a collection of keys
Restore an entire collection of keys from a backup using `beginRestore`.

```java
SyncPoller<KeyVaultBackupOperation, String restorePoller =
    backupClient.beginRestore("<blob-storage-url-to-backup>", "<sas-token>", "<folder-name>");

restorePoller.waitForCompletion();
String blobUri = restorePoller.getFinalResult();

System.out.printf("Backup restored from: %s%n", blobUri);
```

### Selectively restore a key

> Restore an entire collection of keys from a backup using `beginSelectiveRestore`.

Reviewer 2: ```suggestion
Restore an specific key from a backup using `beginSelectiveRestore`.
```

```java
SyncPoller<KeyVaultBackupOperation, String> selectiveRestorePoller =
    backupClient.beginSelectiveRestore("<name-of-key-to-restore>", "<blob-storage-url-to-backup>", "<sas-token>",
        "<folder-name>");

selectiveRestorePoller.waitForCompletion();
String blobUri = selectiveRestorePoller.getFinalResult();

System.out.printf("Key backup restored from: %s%n", blobUri);
```

### Async API

> The following sections provide several code snippets covering some of the most common asynchronous Azure Key Vault Backup service tasks, including:

Reviewer 2: ```suggestion
The following sections provide several code snippets covering some of the most common asynchronous Azure Key Vault Backup client tasks, including:
```

            System.out.println("---------------------------------------------------------------------------------");
            System.out.println(pollResponse.getStatus());
            System.out.println(pollResponse.getValue().getStatus());
            System.out.println(pollResponse.getValue().getStatusDetails());
        }, errorResponse -> {
            System.out.println("---------------------------------------------------------------------------------");
            System.out.println("Error found: " + errorResponse.getMessage());
            errorResponse.printStackTrace();
        }, () -> {
            System.out.println("---------------------------------------------------------------------------------");
            System.out.println("Selective restore operation complete");
        });
```

## Troubleshooting
### General
Azure Key Vault Access Control clients raise exceptions. For example, if you try to retrieve a role assignment after it is deleted a `404` error is returned, indicating the resource was not found. In the following snippet, the error is handled gracefully by catching the exception and displaying additional information about the error.

```java
try {

>     backupClient.getRoleAssignment(KeyVaultRoleAssignmentScope.GLOBAL, "<deleted-role-assginment-name>")

Reviewer 2: ```suggestion
    client.getRoleAssignment(KeyVaultRoleAssignmentScope.GLOBAL, "<deleted-role-assginment-name>")
```
